### ISSUE TITLE
Improve the situation for `no-name-in-module` for C extensions and modules using dynamic features

### ISSUE THREAD

**Issue by chrisspen** [pylint-dev/pylint#1524]
I have the [ProxyTypes](https://pypi.python.org/pypi/ProxyTypes) package installed and have a file `mymodule.py` that imports it like:

    from peak.util.proxies import CallbackProxy

However, pylint throws this false error for it:

    E:  8, 0: No name 'util' in module 'mymodule.peak' (no-name-in-module)

Pylint should reference the `peak` package installed into my virtualenv, and not assume it's a local non-existent module.

**Comment by PCManticore** [comment]
Please provide a minimal working example, I'm not sure I completely understand what is happening, especially since the error you posted indicate something more complex going on than the first import.

**Comment by chrisspen** [comment]
You can easily reproduce it with the following procedure:

    $ mkdir test
    $ cd test
    $ virtualenv .env
    Running virtualenv with interpreter /usr/bin/python2
    New python executable in test/.env/bin/python2
    Also creating executable in test/.env/bin/python
    Installing setuptools, pkg_resources, pip, wheel...done.
    $ .env/bin/pip install pylint ProxyTypes
    Collecting pylint
      Using cached pylint-1.7.1-py2.py3-none-any.whl
    Collecting ProxyTypes
      Downloading ProxyTypes-0.9.zip
    Collecting six (from pylint)
      Using cached six-1.10.0-py2.py3-none-any.whl
    Collecting mccabe (from pylint)
      Using cached mccabe-0.6.1-py2.py3-none-any.whl
    Collecting singledispatch; python_version < "3.4" (from pylint)
      Using cached singledispatch-3.4.0.3-py2.py3-none-any.whl
    Collecting isort>=4.2.5 (from pylint)
      Using cached isort-4.2.15-py2.py3-none-any.whl
    Collecting astroid>=1.5.1 (from pylint)
      Using cached astroid-1.5.3-py2.py3-none-any.whl
    Collecting configparser; python_version == "2.7" (from pylint)
    Collecting backports.functools-lru-cache; python_version == "2.7" (from pylint)
      Using cached backports.functools_lru_cache-1.4-py2.py3-none-any.whl
    Collecting enum34>=1.1.3; python_version < "3.4" (from astroid>=1.5.1->pylint)
      Using cached enum34-1.1.6-py2-none-any.whl
    Collecting lazy-object-proxy (from astroid>=1.5.1->pylint)
      Using cached lazy_object_proxy-1.3.1-cp27-cp27mu-manylinux1_x86_64.whl
    Collecting wrapt (from astroid>=1.5.1->pylint)
    Building wheels for collected packages: ProxyTypes
      Running setup.py bdist_wheel for ProxyTypes ... done
      Stored in directory: /home/chris/.cache/pip/wheels/e3/8f/4c/5d7b9f8e251cc9567b730e7f9e2f74a9ea36a4d9189e0d2864
    Successfully built ProxyTypes
    Installing collected packages: six, mccabe, singledispatch, isort, enum34, lazy-object-proxy, wrapt, backports.functools-lru-cache, astroid, configparser, pylint, ProxyTypes
    Successfully installed ProxyTypes-0.9 astroid-1.5.3 backports.functools-lru-cache-1.4 configparser-3.5.0 enum34-1.1.6 isort-4.2.15 lazy-object-proxy-1.3.1 mccabe-0.6.1 pylint-1.7.1 singledispatch-3.4.0.3 six-1.10.0 wrapt-1.10.10
    $ echo -e "from peak.util.proxies import CallbackProxy\n" > mymodule.py
    $ ls -lah
    total 16K
    drwxrwxr-x  3 chris chris 4.0K Jun 20 16:54 .
    drwxr-xr-x 11 chris chris 4.0K Jun 20 16:52 ..
    drwxrwxr-x  7 chris chris 4.0K Jun 20 16:53 .env
    -rw-rw-r--  1 chris chris   45 Jun 20 16:54 mymodule.py
    $ pylint mymodule.py
    No config file found, using default configuration
    ************* Module mymodule
    C:  2, 0: Trailing newlines (trailing-newlines)
    C:  1, 0: Missing module docstring (missing-docstring)
    E:  1, 0: Unable to import 'peak.util.proxies' (import-error)
    W:  1, 0: Unused CallbackProxy imported from peak.util.proxies (unused-import)


    Report
    ======
    1 statements analysed.

    Statistics by type
    ------------------

    +---------+-------+-----------+-----------+------------+---------+
    |type     |number |old number |difference |%documented |%badname |
    +=========+=======+===========+===========+============+=========+
    |module   |1      |NC         |NC         |0.00        |0.00     |
    +---------+-------+-----------+-----------+------------+---------+
    |class    |0      |NC         |NC         |0           |0        |
    +---------+-------+-----------+-----------+------------+---------+
    |method   |0      |NC         |NC         |0           |0        |
    +---------+-------+-----------+-----------+------------+---------+
    |function |0      |NC         |NC         |0           |0        |
    +---------+-------+-----------+-----------+------------+---------+



    Raw metrics
    -----------

    +----------+-------+------+---------+-----------+
    |type      |number |%     |previous |difference |
    +==========+=======+======+=========+===========+
    |code      |1      |33.33 |NC       |NC         |
    +----------+-------+------+---------+-----------+
    |docstring |0      |0.00  |NC       |NC         |
    +----------+-------+------+---------+-----------+
    |comment   |0      |0.00  |NC       |NC         |
    +----------+-------+------+---------+-----------+
    |empty     |2      |66.67 |NC       |NC         |
    +----------+-------+------+---------+-----------+



    Duplication
    -----------

    +-------------------------+------+---------+-----------+
    |                         |now   |previous |difference |
    +=========================+======+=========+===========+
    |nb duplicated lines      |0     |NC       |NC         |
    +-------------------------+------+---------+-----------+
    |percent duplicated lines |0.000 |NC       |NC         |
    +-------------------------+------+---------+-----------+



    Messages by category
    --------------------

    +-----------+-------+---------+-----------+
    |type       |number |previous |difference |
    +===========+=======+=========+===========+
    |convention |2      |NC       |NC         |
    +-----------+-------+---------+-----------+
    |refactor   |0      |NC       |NC         |
    +-----------+-------+---------+-----------+
    |warning    |1      |NC       |NC         |
    +-----------+-------+---------+-----------+
    |error      |1      |NC       |NC         |
    +-----------+-------+---------+-----------+



    Messages
    --------

    +------------------+------------+
    |message id        |occurrences |
    +==================+============+
    |unused-import     |1           |
    +------------------+------------+
    |trailing-newlines |1           |
    +------------------+------------+
    |missing-docstring |1           |
    +------------------+------------+
    |import-error      |1           |
    +------------------+------------+



    Global evaluation
    -----------------
    Your code has been rated at -70.00/10

    $ .env/bin/python
    Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
    [GCC 5.4.0 20160609] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >>> from peak.util.proxies import CallbackProxy
    >>>

**Comment by meniluca** [comment]
I see the same behaviour with another library. Any idea how to fix this?

**Comment by andrewxiwu** [comment]
Same problem here. Any idea/reference to fix this issue? It is really annoying to have these false positives.

**Comment by neocogent** [comment]
Same problem here with gevent.ssl. 
pylint says SSLContext and PROTOCOL_SSLv23 don't exist but they clearly do. 

Python 2.7.12. gevent 1.2.2

```
python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from gevent.ssl import SSLContext, PROTOCOL_SSLv23
>>> 

```

**Comment by soundlake** [comment]
It seems duplicated. https://github.com/PyCQA/pylint/issues/1138

**Comment by soundlake** [comment]
@neocogent `--unsafe-load-any-extension=y` worked for me.

**Comment by rawrgulmuffins** [comment]
I'm running into a similar issue at my work with a pure python sub-package.

**Comment by jcezarms** [comment]
I'm an year late, but still seeing the exact same behaviour with `tensorflow.python.lib.io`, even though both the library and its `__init__.py` file exist.

As I'm using vscode, this issue causes workspace errors to render in a misleading and incorrect way. To quickly reproduce, install `tensorflow` and type `from tensorflow.python.lib.io import file_io` in a newly created file:
![import-error](https://user-images.githubusercontent.com/25608481/42735380-17098f64-8829-11e8-94cb-dd751863150c.png)

**Comment by brycepg** [comment]
peak issue is due to #1361 
`gevent.ssl` is doing `copy_globals(_source, globals())` (wontfix / maybe astroid brain fix)
tensorflow doesn't directly import the `python` subpackage but one of its imports does which is technically valid python but also kind of fragile. (might be a good issue)

**Comment by rafaelmendy** [comment]
Same problem in importing PyQt5 libraries.
![image](https://user-images.githubusercontent.com/16749759/43641272-39d93882-9738-11e8-8a48-a372ec3e0de7.png)
The code works but pylint send errors.

**Comment by thilux** [comment]
I have the same problem with a file in the same package.

I have the following structure:
project/
        __init__.py
        main.py
        test_project.py


From test_project.py I try importing methods from main doing:

from main import method1, method2

And pyling complains method1 and method2 are not present on main.

Any ideas?

**Comment by eirikgje** [comment]
Another datapoint here:
          
`from netCDF4 import Dataset`
leads to
`[pylint] E0611: No name 'Dataset' in module 'netCDF4' (no-name-in-module)`

**Comment by PCManticore** [comment]
Folks, before commenting here, do make sure to do the following:
- try running pylint first with `pylint --extension-pkg-whitelist=<the package that gives you trouble>`. e.g `pylint --extension-pkg-whitelist=netCDF4 <your files>` This way we can make sure it's not dealing with a C extension package that `pylint` cannot understand
- if that didn't work, it's most likely something related to how that package is structured, in which case we'll need separate issues per each package, as each own can have a different particularity (e.g some of them are namespace packages, some of them uses `globals()`, others import from an inner C extension etc). Creating a separate issue in this case means we don't conflate a problem in a package with another one for this issue.

**Comment by rawrgulmuffins** [comment]
I just tested `--extension-pkg-whistlist` in my case and it did not solve the prolbem.

I'll open an issue for mine but it's unfortunately not an open source project causing this issue so I won't be able to point to a specific project link. =(

**Comment by gaziqbal** [comment]
Is there any guidance on how to bypass the tensorflow.python issue mentioned by @jcezarms above? i'd like to avoid disabling the no-name-in-module error just to get past this false positive.

**Comment by chm123** [comment]
> Is there any guidance on how to bypass the tensorflow.python issue mentioned by @jcezarms above? i'd like to avoid disabling the no-name-in-module error just to get past this false positive.

In VSCode, I had to set
```
    "python.linting.pylintArgs": [
        "--ignored-modules=tensorflow",
        "--extension-pkg-whitelist=tensorflow"
    ],
```
to get rid of the false positives.

**Comment by danielbraun89** [comment]
any updates?

**Comment by dtklein7** [comment]
I'm interested in following this issue as well. I'm getting in VSCode `no-name-in-module` for 2 names in one module, but _not_ the other third function name from that module.

**Comment by ynarwal** [comment]
It happens with specific names of python modules, not sure why though.
I renamed module `factories` to `model_factories` warning went away.

**Comment by ngojason9** [comment]
I created `utils/__init__.py` and it worked. More details can be found on [this Stack Overflow thread](https://stackoverflow.com/questions/40334643/no-name-in-module-error-from-pylint).

**Comment by Jack-CHS-PWD** [comment]
also happens with 

'from win32api import GetShortPathName'
which throws
'No name 'GetShortPathName' in module 'win32api' pylint(no-name-in-module)'

getshortpathname is called later in the code and does work correctly.

**Comment by PCManticore** [comment]
This error usually happens because the given module is either a C extension, so a static analysis tool can't read that source code to figure out the members, or the module tries to be too smart by using dynamic features such as `globals().update` to update the global of one or more files. `pylint` can't really help with these kind of modules, and there are some workarounds that you can use:
- use `--ignored-modules=` to ignore the module that is giving you the error
- use `--generated-members` to specify runtime specific attributes
- use `--extension-pkg-whitelist=` to let `pylint` know that it should import a C extension to build the AST from the live module

**Comment by micklat** [comment]
A suggestion to others who encounter this: ensure that your LD_LIBRARY_PATH (or similar) contains paths to all the dependencies of the module you're trying to import. I had failures like the above when importing a C extension that depended on other extensions, and fixing LD_LIBRARY_PATH apparently made the problem go away.

**Comment by kevlarr** [comment]
I'm running into this issue (and several others) but only on our build machine. Both use pylint `2.4.4`.

Given...

```python
from pydantic import BaseModel
```

Locally (MacOS Mojave 10.14.6) I see no errors.

However, on our build machine (Ubuntu 18.04.1 LTS) I see:

```
E0611: No name 'BaseModel' in module 'pydantic' (no-name-in-module)
```

**Comment by jaysonsantos** [comment]
@kevlarr on your ubuntu it probably does not work because pydantic's wheel ships with a compiled version of the module, which uses cython

**Comment by djelekar** [comment]
`from pydantic import Json`

`E0611: No name 'Json' in module 'pydantic' (no-name-in-module)`

The same thing happens here. What can we do about it?

**Comment by lesinigo** [comment]
Same happens with `setproctitle`, at least since dvarrazzo/py-setproctitle#31 (early 2014).

AFAIK it is reproducible on every kind of system, but if you want a simple test scenario you could get one with this:
```
$ sudo docker run -ti --rm alpine:3.12
... here we are inside the container ...
/ # apk add py3-pip py3-setproctitle
... lots of packages installed ...
/ # pip install pylint
... pylint installed (it's currently not included in Alpine 3.12 packages) ...
/ # pylint --version
pylint 2.5.3
astroid 2.4.2
Python 3.8.5 (default, Jul 20 2020, 23:11:29)
/ # pip list -v | grep setproctitle
setproctitle      1.1.10     /usr/lib/python3.8/site-packages

/ # echo 'from setproctitle import setproctitle' > example.py
/ # pylint -E example.py
************* Module example
/example.py:1:0: E0611: No name 'setproctitle' in module 'setproctitle' (no-name-in-module)
```

Installing `setproctitle` through pip instead of apk does not make any difference:
```
$ sudo docker run -ti --rm alpine:3.12
/ # apk add gcc musl-dev python3-dev py3-pip
/ # pip install pylint setproctitle
/ # pip list -v | grep setproctitle
setproctitle      1.1.10     /usr/lib/python3.8/site-packages
/ # pylint --version
pylint 2.5.3
astroid 2.4.2
Python 3.8.5 (default, Jul 20 2020, 23:11:29)

/ # echo 'from setproctitle import setproctitle' > example.py
/ # pylint -E example.py
************* Module example
/example.py:1:0: E0611: No name 'setproctitle' in module 'setproctitle' (no-name-in-module)
```

**Comment by craigtrim** [comment]
@djelekar same situation.  If you discover a better solution than just putting
`# pylint: disable=no-name-in-module`
on each file with this import, I'd like to hear it!

**Comment by amandalishus** [comment]
This problem also occurs when using Python with Apache Spark via the `pyspark` module, at least when using `pip install pyspark==2.4.6`.  When we started this project using an earlier version of the module the problem did not exist, but we began seeing it after updating to the latest version to stay in sync with AWS Glue.

To reproduce, just install the module as above and import any functions from `pyspark.sql.functions`. For instance:
`from pyspark.sql.functions import col, lit, size`

**Comment by adam-grant-hendry** [comment]
Adding my two cents here:

From the [`pylint 2.15.0-a0` docs on the `no-member / E1101`](https://pylint.pycqa.org/en/latest/user_guide/messages/error/no-member.html?highlight=%22c%20extension%22#no-member-e1101):

> Linting C extension modules is not supported out of the box, especially since pylint has no way to get an AST object out of the extension module.
>
> ...Pylint has a flag, called `extension-pkg-allow-list` (formerly `extension-pkg-whitelist`), through which you can tell it to import that module and to build an AST from that imported module
> ...
> Be aware though that using this flag means that extensions are loaded into the active Python interpreter and may run arbitrary code, which you may not want. This is the reason why we disable by default loading C extensions. In case you do not want the hassle of passing C extensions module with this flag all the time, you can enable `unsafe-load-any-extension` in your configuration file, which will build AST objects from all the C extensions that pylint encounters

So it seems `extension-pkg-allow-list` was made precisely to handle this problem. 

Note that, if you are using static type checking, you can still [create `.pyi` stubs ](https://mypy.readthedocs.io/en/stable/stubgen.html?highlight=extension%20modules#specifying-how-to-generate-stubs) for static analysis. (If you are making a package, be sure to include the requisite `py.typed` marker (file) to the root of the package (per [PEP 561](https://mypy.readthedocs.io/en/latest/installed_packages.html#creating-pep-561-compatible-packages)).

Interestingly, when `extension-pkg-allow-list` isn't set, if stubs for your C extension module use wildcard imports (i.e. `from package import *`), you will see a [`no-name-in-module / E0611`](https://pylint.pycqa.org/en/latest/user_guide/messages/error/no-name-in-module.html?highlight=E0611#no-name-in-module-e0611)error, which is created by the [`pylint` `variables` checker](https://github.com/PyCQA/pylint/blob/main/pylint/checkers/variables.py) rather than a [`c-extension-no-member / I1101`](https://pylint.pycqa.org/en/latest/user_guide/messages/information/c-extension-no-member.html?highlight=extension%20modules#c-extension-no-member-i1101), which is created by the [`pylint` `typechecker` checker](https://github.com/PyCQA/pylint/blob/main/pylint/checkers/typecheck.py).

Although `mypy` allows [using wildcard imports to export modules](https://peps.python.org/pep-0484/#stub-files), this last one may surprise you. I've seen this happen occasionally when [the name you import has been imported in the target module with a wildcard](https://pycodequ.al/docs/pylint-messages/e0611-no-name-in-module.html). Rest assured, your extension module is typed, but just be sure to `extension-pkg-allow-list` (and only set it for packages you trust!).

**Comment by jace** [comment]
> So it seems `extension-pkg-allow-list` was made precisely to handle this problem.
> 
> Note that, if you are using static type checking, you can still [create `.pyi` stubs ](https://mypy.readthedocs.io/en/stable/stubgen.html?highlight=extension%20modules#specifying-how-to-generate-stubs) for static analysis. (If you are making a package, be sure to include the requisite `py.typed` marker (file) to the root of the package (per [PEP 561](https://mypy.readthedocs.io/en/latest/installed_packages.html#creating-pep-561-compatible-packages)).)

How do you get pylint to read a `.pyi` stub file instead of the regular `.py` file? I'm using [mkinit](https://github.com/Erotemic/mkinit) to create a lazy-loading `__init__.py` with a corresponding `__init__.pyi` stub file. Type checkers are fine with this, but Pylint is not. I have new false positives for `isinstance-second-argument-not-valid-type`, `no-name-in-module` and `not-callable`.

**Comment by jacobtylerwalls** [comment]
Thanks @adam-grant-hendry for the write up! Closing as resolved. pyi support tracked separately.
